src/modm/platform/dma/stm32/dma_hal.hpp.in

```diff
@@ -0,0 +1,304 @@
+				uint32_t(circular);
+		direction = dir;
+	}
+
+	/**
+	 * Start the transfer of the DMA channel
+	 */
+	static void
+	start();
+	/**
+	 * Stop a DMA channel transfer
+	 */
+	static void
+	stop();
+
+	/**
+	 * Get the direction of the data transfer
+	 */
+	static DataTransferDirection
+	getDataTransferDirection();
+
+	/**
+	 * Set the source and destination address of the DMA channel
+	 *
+	 * Requires to set the direction of the DMA channel first
+	 *
+	 * @param[in] sourceAddress Source address of DMA transfer
+	 * @param[in] destinationAddress Destination address of DMA transfer
+	 */
+	static void
+	setAddresses(uint32_t sourceAddress, uint32_t destinationAddress)
+	{
+		modm_assert_continue_ignore_debug(uint32_t(direction) != 0xff,
+				"dma", "setAddresses", "direction not configured");
+
+		DMA_Channel_TypeDef *Base = (DMA_Channel_TypeDef *) CHANNEL_BASE;
+		if (direction == DataTransferDirection::MemoryToPeripheral) {
+			Base->CMAR = sourceAddress;
+			Base->CPAR = destinationAddress;
+		} else {
+			Base->CPAR = sourceAddress;
+			Base->CMAR = destinationAddress;
+		}
+	}
+	/**
+	 * Set the source address of the DMA channel
+	 *
+	 * @param[in] address Source address
+	 */
+	static void
+	setSourceAddress(uint32_t address)
+	{
+		modm_assert_continue_ignore_debug(uint32_t(direction) != 0xff,
+				"dma", "setSourceAddress", "direction not configured");
+
+		DMA_Channel_TypeDef *Base = (DMA_Channel_TypeDef *) CHANNEL_BASE;
+		if (direction == DataTransferDirection::MemoryToPeripheral)
+			Base->CMAR = address;
+		else
+			Base->CPAR = address;
+	}
+	/**
+	 * Set destination address of DMA channel
+	 *
+	 * @param[in] address Destination address
+	 */
+	static void
+	setDestinationAddress(uint32_t address)
+	{
+		modm_assert_continue_ignore_debug(uint32_t(direction) != 0xff,
+				"dma", "setDestinationAddress", "direction not configured");
+
+		DMA_Channel_TypeDef *Base = (DMA_Channel_TypeDef *) CHANNEL_BASE;
+		if (direction == DataTransferDirection::MemoryToPeripheral)
+			Base->CPAR = address;
+		else
+			Base->CMAR = address;
+	}
+
+	/**
+	 * Enable/disable memory increment
+	 *
+	 * When enabled, the memory address is incremented by the size of the data
+	 * (e.g. 1 for byte transfers, 4 for word transfers) after the transfer
+	 * completed.
+	 *
+	 * @param[in] increment Enable/disable
+	 */
+	static void
+	setMemoryIncrementMode(bool increment)
+	{
+		DMA_Channel_TypeDef *Base = (DMA_Channel_TypeDef *) CHANNEL_BASE;
+		if (increment)
+			Base->CCR |= uint32_t(MemoryIncrementMode::Increment);
+		else
+			Base->CCR &= ~uint32_t(MemoryIncrementMode::Increment);
+	}
+
+	/**
+	 * Enable/disable peripheral increment
+	 *
+	 * When enabled, the peripheral address is incremented by the size of the data
+	 * (e.g. 1 for byte transfers, 4 for word transfers) after the transfer
+	 * completed.
+	 *
+	 * @param[in] increment Enable/disable
+	 */
+	static void
+	setPeripheralIncrementMode(bool increment)
```

Comment:
I'm not sure if I would prefer a less hardware-specific abstraction (`setSourceIncrementMode(...)` / `setDestinationIncrementMode(...)`), even if this would add complexity here...

Reply:
See my comment [here](https://github.com/modm-io/modm/pull/371#issuecomment-621647643).